### PR TITLE
Feature/pattern node types

### PR DESCRIPTION
**Pull request by lharzenetter**
- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request. Especially, we require **a single commit**
- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Ensure that you appear in `NOTICE` at Copyright Holders
- [ ] Tests created for changes
- [x] Screenshots added (for UI changes)

Pattern Node Types are now rendered in a different view in the Topology Modeler.

grafik

grafik

### CODE REVIEW

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/apiData/NodeTypesVisualsApiData.java [hunk 1]
```diff
@@ -14,26 +14,50 @@
...
     public String color;
+    public String pattern;
     public QName nodeTypeId;
 
     public NodeTypesVisualsApiData(VisualAppearanceResource visuals) {
+        NodeTypeId parent = (NodeTypeId) visuals.getId().getParent();
+        TNodeType element = RepositoryFactory.getRepository().getElement(parent);
+
         this.color = visuals.getBorderColor();
-        this.nodeTypeId = ((NodeTypeId) visuals.getId().getParent()).getQName();
```
Comment (koppor) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/apiData/NodeTypesVisualsApiData.java]: Why was it required to change this code? I would have assumed that it works?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/apiData/NodeTypesVisualsApiData.java [hunk 2]
```diff
@@ -14,26 +14,50 @@
...
+            Optional<TTag> isPattern = element.getTags()
+                .getTag()
+                .stream()
+                .filter(
+                    tTag -> "pattern".equals(
+                        tTag.getName().toLowerCase()
+                    )
+                )
+                .findFirst();
+
```
Comment (koppor) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/apiData/NodeTypesVisualsApiData.java]: Just chain.

```java
.findFirst()
.ifPresent...
```

**Comment by lharzenetter** [comment]
I implemented the proposed approach using Pattern Namespaces, as shown in the added screenshot.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/apiData/NodeTypesVisualsApiData.java [hunk 3]
```diff
@@ -14,34 +14,43 @@
 
 package org.eclipse.winery.repository.rest.resources.apiData;
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/apiData/NodeTypesVisualsApiData.java]: Change copyright year ^^

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytypes/properties/PropertiesDefinitionResource.java [hunk 4]
```diff
@@ -138,7 +138,7 @@ public Response onJsonPost(PropertiesDefinitionResourceApiData data) {
             ModelUtilities.replaceWinerysPropertiesDefinition(et, data.winerysPropertiesDefinition);
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/entitytypes/properties/PropertiesDefinitionResource.java]: change copyright year ^^

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/resources/admin/NamespaceResourceTest.java [hunk 5]
```diff
@@ -25,4 +25,9 @@ public void getNamespaceList() throws Exception {
         this.assertGet("admin/namespaces/?all", "entitytypes/admin/namspacesList.json");
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/resources/admin/NamespaceResourceTest.java]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/instance/admin/namespaces/namespaces.component.html [hunk 6]
```diff
@@ -18,7 +18,9 @@
     <winery-table title="Defined Prefixes for Namespaces"
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/instance/admin/namespaces/namespaces.component.html]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/instance/admin/namespaces/namespaces.component.ts [hunk 7]
```diff
@@ -11,16 +11,15 @@
  *
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/instance/admin/namespaces/namespaces.component.ts]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/instance/admin/namespaces/namespaces.service.ts [hunk 8]
```diff
@@ -16,7 +16,7 @@ import { WineryNamespaceSelectorService } from '../../../wineryNamespaceSelector
 import { Observable } from 'rxjs';
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/instance/admin/namespaces/namespaces.service.ts]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/model/namespaceProperties.ts [hunk 9]
```diff
@@ -12,8 +12,13 @@
  * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/model/namespaceProperties.ts]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/wineryNamespaceSelector/wineryNamespaceSelector.component.ts [hunk 10]
```diff
@@ -15,7 +15,7 @@ import {Component, ElementRef, forwardRef, Input, OnInit, ViewChild} from '@angu
 import {WineryNamespaceSelectorService} from './wineryNamespaceSelector.service';
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/wineryNamespaceSelector/wineryNamespaceSelector.component.ts]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/wineryNamespaceSelector/wineryNamespaceSelector.service.ts [hunk 11]
```diff
@@ -14,7 +14,7 @@
 import { Injectable } from '@angular/core';
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/wineryNamespaceSelector/wineryNamespaceSelector.service.ts]: copyright year ^^

File: org.eclipse.winery.repository.ui/src/app/wineryRepository.component.ts [hunk 12]
```diff
@@ -62,7 +62,7 @@ export class WineryRepositoryComponent implements OnInit {
     }
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository.ui/src/app/wineryRepository.component.ts]: copyright year ^^

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/Constants.java [hunk 13]
```diff
@@ -46,7 +46,6 @@
     // suffix for ZIP files
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/Constants.java]: copyright year

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/filebased/AutoSaveListener.java [hunk 14]
```diff
@@ -59,14 +59,14 @@ public void configurationChanged(ConfigurationEvent event) {
                     Files.createDirectories(this.path.getParent());
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/filebased/AutoSaveListener.java]: copyright year ^^

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/filebased/NamespaceProperties.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.Serializable;
+
+public class NamespaceProperties implements Serializable, Comparable<NamespaceProperties> {
+
+    private static final long serialVersionUID = -6867642303902116547L;
+
+    private String namespace;
+    private String prefix;
+    private String readableName = "";
+    private String upstreamRepository = "";
```
Comment (ghareeb-falazi) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/filebased/NamespaceProperties.java]: add a comment explaining the purpose of the `upstreamRepository` field as it is not clear from the context of the pull requrest

**Comment by lharzenetter** [comment]
@ghareeb-falazi I adjusted all the license headers

**Pull request merged**